source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1632 @@
+                {
+                  // + or - topography based on the initial noise magnitude.
+                  const double h_seed = (std::rand()%( 2*noise_h+1 )) - noise_h;
+                  h[i] = h[i] + h_seed;
+                }
+
+              // Here we add the sediment rain (m/yr) as a flat increase in height.
+              // This is done because adding it as an uplift rate would affect the basement.
+              if (sediment_rain > 0 && use_marine)
+                {
+                  // Only apply sediment rain to areas below sea level.
+                  if (h[i] < sl)
+                    {
+                      // If the rain would put us above sea level, set height to sea level.
+                      if (h[i] + sediment_rain*a_dt > sl)
+                        h[i] = sl;
+                      else
+                        h[i] = std::min(sl,h[i] + sediment_rain*a_dt);
+                    }
+                }
+            }
+
+          // The ghost nodes are added as a single layer of points surrounding the entire model.
+          // For example, if ASPECT's surface mesh is a 2D surface that is 3x3 (nx x ny) points,
+          // FastScape will be set as a 2D 5x5 point surface. On return to ASPECT, the outer ghost nodes
+          // will be ignored, and ASPECT will see only the inner 3x3 surface of FastScape.
+          if (use_ghost_nodes)
+            set_ghost_nodes(h.get(), vx.get(), vy.get(), vz.get());
+
+          // If specified, apply the orographic controls to the FastScape model.
+          if (use_orographic_controls)
+            apply_orographic_controls(h.get(), kd.get(), kf.get());
+
+          // Get current FastScape timestep.
+          int istep = 0;
+          fastscape_get_step_(&istep);
+
+          // Set velocity components.
+          if (use_velocities)
+            {
+              fastscape_set_u_(vz.get());
+              fastscape_set_v_(vx.get(), vy.get());
+            }
+
+          // Set h to new values, and erosional parameters if there have been changes.
+          fastscape_set_h_(h.get());
+          fastscape_set_erosional_parameters_(kf.get(), &kfsed, &m, &n, kd.get(), &kdsed, &g, &g, &p);
+
+          // Find  timestep size, run fastscape, and make visualizations.
+          execute_fastscape(h.get(), kd.get(), istep);
+
+          // Write a file to store h, b & step for restarting.
+          // TODO: It would be good to roll this into the general ASPECT checkpointing,
+          // and when we do this needs to be changed.
+          if (((this->get_parameters().checkpoint_time_secs == 0) &&
+               (this->get_parameters().checkpoint_steps > 0) &&
+               ((current_timestep + 1) % this->get_parameters().checkpoint_steps == 0)) ||
+              (this->get_time() + a_dt >= end_time && this->get_timestepping_manager().need_checkpoint_on_terminate()))
+            {
+              save_restart_files(h.get(), b.get(), sf.get(), istep);
+            }
```

Comment:
There are two ways to connect this into the general checkpointing mechanism:
1. Create a checkpointing architecture with `serialize` function like we have for the postprocessors (see, e.g. here https://github.com/geodynamics/aspect/blob/18d75ebaaab2d409685b87633701de676c00720f/source/postprocess/visualization.cc#L314). This would allow you to store all the data as part of ASPECT's main checkpoint files. However, this architecture does not exist yet for mesh deformation plugins and would need a bit of work.
2. Make use of the `pre_checkpoint_store_user_data` signal that is triggered every time before a checkpoint is written. You can see how this signal is used here: https://github.com/geodynamics/aspect/blob/18d75ebaaab2d409685b87633701de676c00720f/source/particle/world.cc#L218. I think this is the easier approach. To make this work, you would move all commands that need to be executed on checkpointing into a separate function (`prepare_checkpoint`) and then connect this function to the signal in the `initialize` function like this:
```
this->get_signals().pre_checkpoint_store_user_data.connect(
            [&] (typename parallel::distributed::Triangulation<dim> &)
          {
            prepare_checkpoint();
          });
```
 
Once connected the `prepare_checkpoint` function will be executed every time a checkpoint is written.


Similarly you could use the `post_resume_load_user_data` signal to load data after restarting from a checkpoint.

Reply:
I tried to use the pre_checkpoint_store_user_data, but I seem to be running into a problem where by the time the signal is called, the h data that exists only on processor 0 during the compute_velocity_constraints_on_boundary is no longer accessible. I think a way around this could be to add 3 mutable vectors to fastscape.h, and then resize  these to array_size during initialize and update them with the relevant data while still on processor zero within the function. Does this seem like an okay way to do this, or am I missing an easier way to do it?